### ISSUE TITLE
Individual localContexts for entries in List

### ISSUE THREAD

**Issue by SimonSchneider** [graphql-java/graphql-java#1559]
The localContext concept introduced in v12 allows us to send a localContext to child nodes. This gives a proper solution to the antipattern of adding "look ahead data" to the DTOs or having multiple DTOs with varying amounts of fields on them to pass information to child nodes.

A possible extension of this feature could be to allow returning a List of `DataFetcherResults` and allow each entry in the list to have its own localContext.

The justification is that often when you return a List of DTOs they will all have their own conceptual context and not supporting functionality to return individual LocalContexts for each of them will force usage of the above-mentioned antipattern.

I would imagine an API like this:
```java
@PublicSpi
public interface DataFetcher<T> {
    

    /**
     * This is called by the graphql engine to fetch the value.  The {@link graphql.schema.DataFetchingEnvironment} is a composite
     * context object that tells you all you need to know about how to fetch a data value in graphql type terms.
     *
     * @param environment this is the data fetching environment which contains all the context you need to fetch a value
     *
     * @return a value of type T. May be wrapped in a {@link graphql.execution.DataFetcherResult} or in a List of {@link graphql.execution.DataFetcherResult}
     *
     * @throws Exception to relieve the implementations from having to wrap checked exceptions. Any exception thrown
     *                   from a {@code DataFetcher} will eventually be handled by the registered {@link graphql.execution.DataFetcherExceptionHandler}
     *                   and the related field will have a value of {@code null} in the result.
     */
    T get(DataFetchingEnvironment environment) throws Exception;
}

```

implementation:
```java
class Test implements DataFetcher<Dto> {
      List<DataFetcherResult<Dto>> get(DataFetchingEnvironment environment) {
          ...
     }
}
```

**Comment by bbakerman** [comment]
This is not a crazy idea.  Its stripes the errors and local context to each return object

Can you go into more details about why you need this.  As in a more specific use case.

**Comment by SimonSchneider** [comment]
Thank you @bbakerman

The use case I have is that we have a structure like this:

```
type User {
  purchases: [Purchases!]!
}

type Purchase {
  id: ID!
  ... # various metadata existing in other systems (could be nested in structures)
  a: A!
}

type A {
  ... # more metadata from other systems
}
```

We have a relationship of a user to its purchases as well as some ids for each purchase that are used to fetch data from other systems. All of these IDs are loaded together when we resolve a user's purchases to improve performance. With the purchaseIDs we create the list of purchases (and currently put a context containing the rest of the related IDs as a member for the Purchase class).
When we return DataFetcherResults from the resolver for purchase fields we set the context of that purchase as the local context for the child.

The extra IDs are not exposed on the actual purchase but are simply used to fetch data from upstream systems.
IDs may, for example, be internal IDs for the purchase in the other systems.

Therefore we would like to have each purchase have it's own local context as they are conceptually different entities. And we would not have the context object as a member on the purchase before returning it as the local context once we are resolving purchase fields.

I hope this clears up the use case.

**Comment by dimdimych** [comment]
We could have an option to provide an instance of Function source -> context as a localContext  which would be called for each Data Fetcher Environment, this way we wouldn't introduce a breaking change in DataFetcherResult processing

**Comment by dimdimych** [comment]
Please disregard my comment above, I think @SimonSchneider 's solution is cleaner, I've implemented this approach in #1567
